### PR TITLE
Otel phoenix migration

### PR DESCRIPTION
**Pull request by bryannaegele**
Migrate otel phoenix to here and test setup.

Marking as WIP and not draft so CI runs.

### CODE REVIEW

File: .gitignore
Comment (indrekj) [inline .gitignore:26]: shouldn't this be `.tool-versions`? assuming `asdf` is used

**Comment by ferd** [comment]
On the form this looks fine, but since Tristan is marked as the other codeowner for this I'll leave it to him to approve the merge.

File: .github/test-matrix.json
Comment (indrekj) [inline .github/test-matrix.json:1]: i think this file is unused at the moment

File: .github/workflows/elixir.yml
Comment (indrekj) [inline .github/workflows/elixir.yml:36]: should also 1.10 be included? I remember from the opentelemetry_tesla library that 1.10 was mentioned as the lowest version in the OTEL land.

Probably unimportant, but I think patch version doesn't need to be included, e.g. `elixir_version: ["1.12", "1.11", "1.10"]`

File: instrumentation/opentelemetry_phoenix/lib/opentelemetry_phoenix.ex
Comment (indrekj) [inline instrumentation/opentelemetry_phoenix/lib/opentelemetry_phoenix.ex:157]: unrelated to the migration: I wonder if this should be inside a try/catch block or if there's should be a mention of `Plug.Telemetry` requirement in the README to avoid this cryptic error: https://github.com/opentelemetry-beam/opentelemetry_phoenix/issues/34#issuecomment-905275488

File: instrumentation/opentelemetry_phoenix/mix.exs
Comment (indrekj) [inline instrumentation/opentelemetry_phoenix/mix.exs:59]: Shouldn't this be only `dev/test` dependency? I'd assume only opentelemetry_api is needed for all the calls, and the user could use whichever implementation they choose (though I guess there's only one atm).

File: instrumentation/opentelemetry_phoenix/mix.exs
Comment (indrekj) [inline instrumentation/opentelemetry_phoenix/mix.exs:61]: Could you also add `1.0`, as in `"~> 0.4 or ~> 1.0"`.

Version 1.0 was released without backwards-incompatible changes: https://github.com/beam-telemetry/telemetry/blob/main/CHANGELOG.md#100

File: instrumentation/opentelemetry_phoenix/test/test_helper.exs
Comment (indrekj) [inline instrumentation/opentelemetry_phoenix/test/test_helper.exs:7]: this commented out code can probably be removed?

**Review by indrekj — APPROVED** [review]
Went over the PR as well. Added a couple of nitpick comments, nothing major.

File: .github/test-matrix.json
Comment (bryannaegele) [reply .github/test-matrix.json:1]: It's just unused for the moment to get this merged.

File: .github/workflows/elixir.yml
Comment (bryannaegele) [reply .github/workflows/elixir.yml:36]: 1.10 has its own include in the matrix. It's not compatible with OTP 24.

File: instrumentation/opentelemetry_phoenix/lib/opentelemetry_phoenix.ex
Comment (bryannaegele) [reply instrumentation/opentelemetry_phoenix/lib/opentelemetry_phoenix.ex:157]: Feel free to add a doc PR to note the req for plug telemetry.

**Comment by bryannaegele** [comment]
> On the form this looks fine, but since Tristan is marked as the other codeowner for this I'll leave it to him to approve the merge.

The code of the phoenix project itself wasn't the main aim on this PR. It was just a starting point to get the folder structure laid out and the CI going. :)

Erlang approvers/maintainers have the overall control of the direction of the repo, so you're perfectly ok to add your reviews on anything. The project-level (directory) owners is a way to enable the project owners coming in the ability to do normal maintainer things but restricted to their project.

**Review by ferd — APPROVED** [review]

File: .github/test-matrix.json
Comment (bryannaegele) [reply .github/test-matrix.json:9]: I was trying (unsuccessfully) to get a solution working where we can pull in a config and assign it to the strategy. I know it can be done (there are examples) but I just aborted it temporarily to get things at least working. Until we can get that working, we have to copy the matrix around. Otherwise, it's going to be a maintenance headache.

The `include` part of a matrix strategy means "also add this one combination to the final matrix". In our case, 1.10 isn't compatible with OTP >22 but we do support Elixir 1.10. Basically, just creating a matrix that covers min to max based on https://hexdocs.pm/elixir/1.12/compatibility-and-deprecations.html#compatibility-between-elixir-and-erlang-otp. Conversely, you can tell it to drop some specific set from the matrix with the `exclude`

I do think the exclude can be dropped and is leftover from one of the other projects.

File: .github/workflows/elixir.yml
Comment (bryannaegele) [reply .github/workflows/elixir.yml:25]: https://github.com/open-telemetry/opentelemetry-erlang-contrib/issues/2#issuecomment-910560900 ;)

**Pull request merged**
